File: src/main/java/com/extendedae_plus/client/render/widgets/button/ButtonImplementations.java
package com.extendedae_plus.client.render.widgets.button;

import appeng.api.config.YesNo;
import appeng.client.gui.AEBaseScreen;
import appeng.core.network.serverbound.ConfigButtonPacket;
import com.extendedae_plus.common.init.ModSettings;
import com.extendedae_plus.mixin.impl.bridge.HelperProviderButtons;
import com.extendedae_plus.mixin.impl.bridge.SyncerSmartBlocking;
import com.extendedae_plus.mixin.impl.bridge.SyncerSmartDoubling;
import com.extendedae_plus.mixin.impl.widget.HelperRenderablesModifier;
import com.mojang.datafixers.util.Pair;
import net.minecraft.world.inventory.AbstractContainerMenu;
import org.jetbrains.annotations.Nullable;

public class ButtonImplementations {
    public static EAEPServerCycleButton buttonBlocking(AbstractContainerMenu menu) {
        EAEPServerCycleButton button = new EAEPServerCycleButton.Builder()
                .setTask(new ConfigButtonPacket(ModSettings.smartBlocking, false))
                .addPart(EAEPActionItems.blockingDisabled)
                .addPart(EAEPActionItems.blockingEnabled)
                .addPart(EAEPActionItems.blockingUnable)
                .setIterator((prev, reversed) -> (prev + (reversed ? -1 : 1)) % 2)
                .setSyncer(() -> {
                    if (!(menu instanceof SyncerSmartBlocking syncer)) return 0;
                    else return switch (syncer.eaep$getBlockingState()) {
                        case DISABLED -> 0;
                        case ENABLED -> 1;
                        case DISABLED_BY_SUPER -> 2;
                    };
                })
                .build();
        button.updateState();
        return button;
    }

    public static EAEPServerCycleButton buttonDoubling(AbstractContainerMenu menu) {
        EAEPServerCycleButton button = new EAEPServerCycleButton.Builder()
                .setTask(new ConfigButtonPacket(ModSettings.smartDoubling, false))
                .addPart(EAEPActionItems.doublingDisabled)
                .addPart(EAEPActionItems.doublingEnabled)
                .setSyncer(() -> menu instanceof SyncerSmartDoubling syncer
                        && YesNo.YES.equals(syncer.eaep$getDoublingState()))
                .build();
        button.updateState();
        return button;
    }

    public static <T extends AEBaseScreen<?>> Pair<Integer, Integer>
    updateScalingButtonsLayout(T screen, int bx, int by, @Nullable Pair<Integer, Integer> lastScreenInfo) {
        if (!(screen instanceof HelperProviderButtons helper)) return lastScreenInfo;

        boolean flagReplaceButton = lastScreenInfo == null
                || screen.width != lastScreenInfo.getFirst()
                || screen.height != lastScreenInfo.getSecond();
        if (flagReplaceButton) lastScreenInfo = new Pair<>(screen.width, screen.height);

        int spacing = helper.eaep$getButtons().getFirst().getHeight() + 6;
        helper.eaep$getButtons().forEach(button -> {
            if (button == null) return;
            button.setVisibility(true);
            if (!screen.renderables.contains(button))
                HelperRenderablesModifier.addRenderableWidget(screen, button);

            if (flagReplaceButton) {
                HelperRenderablesModifier.removeWidget(screen, button);
                HelperRenderablesModifier.addRenderableWidget(screen, button);
            }

            button.setX(bx);
            button.setY(by + spacing * helper.eaep$getButtons().indexOf(button));
        });

        return lastScreenInfo;
    }
}

File: src/main/kotlin/com/extendedae_plus/client/render/widgets/button/ButtonImplementations.kt
package com.extendedae_plus.client.render.widgets.button

import appeng.api.config.YesNo
import appeng.client.gui.AEBaseScreen
import appeng.core.network.serverbound.ConfigButtonPacket
import com.extendedae_plus.common.init.EAEPSettings
import com.extendedae_plus.common.registry.settings.StateSmartBlocking
import com.extendedae_plus.mixin.helper.HelperProviderButtons
import com.extendedae_plus.mixin.helper.SyncerSmartBlocking
import com.extendedae_plus.mixin.helper.SyncerSmartDoubling
import com.extendedae_plus.mixin.impl.HelperRenderablesModifier
import com.mojang.datafixers.util.Pair
import net.minecraft.world.inventory.AbstractContainerMenu

object ButtonImplementations {
    @JvmStatic
    fun buttonBlocking(menu: AbstractContainerMenu): EAEPServerCycleButton {
        val button = EAEPServerCycleButton.Builder()
            .setTask(ConfigButtonPacket(EAEPSettings.smartBlocking, false))
            .addPart(EAEPActionItems.BlockingDisabled)
            .addPart(EAEPActionItems.BlockingEnabled)
            .addPart(EAEPActionItems.BlockingUnable)
            .setIterator { prev, reversed -> (prev + (if (reversed) -1 else 1)) % 2 }
            .setSyncer {
                if (menu !is SyncerSmartBlocking) return@setSyncer 0
                else return@setSyncer when (menu.`eaep$getBlockingState`()) {
                    StateSmartBlocking.DISABLED -> 0
                    StateSmartBlocking.ENABLED -> 1
                    StateSmartBlocking.DISABLED_BY_SUPER -> 2
                }
            }
            .build()
        button.updateState()
        return button
    }

    @JvmStatic
    fun buttonDoubling(menu: AbstractContainerMenu): EAEPServerCycleButton {
        val button = EAEPServerCycleButton.Builder()
            .setTask(ConfigButtonPacket(EAEPSettings.smartDoubling, false))
            .addPart(EAEPActionItems.DoublingDisabled)
            .addPart(EAEPActionItems.DoublingEnabled)
            .setSyncer(SyncerBooleanGeneric {
                menu is SyncerSmartDoubling
                        && YesNo.YES == menu.`eaep$getDoublingState`()
            })
            .build()
        button.updateState()
        return button
    }

    @JvmStatic
    fun <T : AEBaseScreen<*>> updateScalingButtonsLayout(
        screen: T,
        bx: Int,
        by: Int,
        avoidToolbox: Boolean,
        lastScreenInfo: Pair<Int, Int>?
    ): Pair<Int, Int> {
        var lastScreenInfo = lastScreenInfo
        if (screen !is HelperProviderButtons) return Pair(0, 0)

        val flagReplaceButton = lastScreenInfo == null
                    || screen.width != lastScreenInfo.first
                    || screen.height != lastScreenInfo.second
        if (flagReplaceButton)
            lastScreenInfo = Pair(screen.width, screen.height)

        val buttons = screen.`eaep$getButtons`()
        val spacing: Int = buttons[0].height + 6
        buttons.forEach { button ->
            button.setVisibility(true)
            if (!screen.renderables.contains(button))
                HelperRenderablesModifier.addRenderableWidget(screen, button)

            if (flagReplaceButton) {
                HelperRenderablesModifier.removeWidget(screen, button)
                HelperRenderablesModifier.addRenderableWidget(screen, button)
            }

            val indexButton = buttons.indexOf(button)
            button.x = bx + (if (avoidToolbox && indexButton >= 3) spacing else 0)
            button.y = by + spacing * (if (avoidToolbox) indexButton % 3 else indexButton)
        }

        return lastScreenInfo
    }
}
